text/todo-generic-payload.md

```diff
@@ -0,0 +1,368 @@
+
+#### Key Features:
+
+- **Type Safety**: Strongly typed value system supporting 6 data types (INT32/UINT32/REAL32/INT64/UINT64/REAL64)
+- **Extensibility**: Extension fields for additional metadata without breaking compatibility
+- **Discovery**: Self-describing payloads that announce their capabilities
+- **Flexibility**: Supports logical, continuous, and discrete function types
+- **Control Modes**: Latching and momentary control with configurable timeouts
+
+### Protocol Operation
+
+1. **Discovery Phase**: Payload broadcasts `GENERIC_PAYLOAD_STATUS` at 1Hz
+2. **Capability Query**: Vehicle requests function/telemetry metadata via `MAV_CMD_REQUEST_MESSAGE`
+3. **Operation**: Vehicle sends control commands, payload streams telemetry and status updates
+
+### Enumerations
+
+#### Control Modes
+
+- `GENERIC_PAYLOAD_CONTROL_MODE`: Single-value enumeration for selecting control mode
+  - `LATCHING` (1): Function maintains state until changed
+  - `MOMENTARY` (2): Function auto-deactivates after timeout
+- `GENERIC_PAYLOAD_CONTROL_MODE_FLAGS`: Bitmask for supported control modes
+  - `SUPPORTS_LATCHING` (1): Function supports latching mode
+  - `SUPPORTS_MOMENTARY` (2): Function supports momentary mode
+
+#### Value Types
+
+Support for 6 data types covering most payload needs:
+
+- 32-bit: INT32, UINT32, REAL32
+- 64-bit: INT64, UINT64, REAL64
+
+#### Function Types
+
+- `LOGICAL`: Boolean controls (on/off, enable/disable)
+- `CONTINUOUS`: Analog controls with min/max ranges (servo angles, speeds)
+- `DISCRETE`: Digital controls with enumerated values (modes, frequencies)
+
+### Message ID Allocation
+
+Proposed message IDs in the 60000-60004 range (vendor-specific range) to avoid conflicts during development. Final standardization would require official MAVLink message ID allocation.
+
+### Extension Fields
+
+All messages include extension fields for future enhancements:
+
+- Power consumption monitoring
+- Temperature sensing
+- Mass/inertia properties
+- 64-bit value support
+
+## Alternatives
+
+### Alternative 1: Extend Existing Payload Commands
+
+Build upon the current [MAVLink payload services](https://mavlink.io/en/services/payload.html) by adding more generic commands and standardizing implementation.
+
+**Analysis**: The existing `MAV_CMD_DO_SET_ACTUATOR` and `MAV_CMD_DO_SET_SERVO` commands could theoretically be extended with additional parameters for type information and metadata.
+
+**Key Limitations**:
+
+- Commands are stateless and don't provide discovery mechanisms
+- Inconsistent implementation across flight stacks ("only implemented on PX4")
+- Limited parameter space in command messages prevents rich metadata
+- No built-in telemetry streaming capabilities
+- Would require significant breaking changes to existing command semantics
+
+### Alternative 2: Parameter-Only Approach
+
+Use MAVLink's parameter system (`PARAM_*` messages) to expose payload functions as parameters.
```

Comment:
MAVLink explicitly doesn't mandate what parameters are used - they are internal. The only exception is some of the [standardized names in the camera definition format](https://mavlink.io/en/services/camera_def.html#common-parameters) which need to be known in order to construct UIs. 

So while this is possible, I'd like to avoid it "directly"

Component metadata provides a mechanism to abstract parameter names - i.e. you can standardise a reference to a parameter name, and the component itself can name the parameter as it likes.